test-tests/shared/src/test/scala/zio/test/ReportingTestUtils.scala

```diff
@@ -156,6 +156,13 @@ object ReportingTestUtils {
     withOffset(2)(assertSourceLocation() + "\n")
   )
 
+  def test9(implicit trace: ZTraceElement): ZSpec[Any, Nothing] = test("labeled failures") {
+    assertTrue(1 == 1).map(_.label("first")) &&
+    assertTrue(1 == 1).map(_.label("second")) &&
+    assertTrue(1 == 0).map(_.label("third")) &&
+    assertTrue(1 == 0).map(_.label("fourth"))
```

Comment:
Probably would be slightly more ergonomic if `label` could be called directly, such as `assertTrue(1 == 1).label("something")`. But the unification of these different assertion result types may make that implementation a bit painful. This is a great start though.